docs/developer_docs/developer_notes.md

```diff
@@ -0,0 +1,192 @@
+  and `pymat_converter` instance, executes the Python code and converts the results to Matlab format.
+
+Finally, there is a Python class, `libpymcr.MatlabProxyObject`
+which overloads various Python dunder methods with equivalent calls to Matlab methods
+to allow the "opaque" Matlab objects to be inspected from Python.
+
+
+## Type conversion
+
+The core of the module is the `libpymcr::pymat_converter` [class](../../src/type_converter.hpp)
+which converts data from Python to Matlab and vice versa, according to:
+
+| Python to    | Matlab   || Matlab to    | Python     |
+--------------------------------------------------------
+| `list`       | `cell`   || `cell`       | `list`     |
+| `tuple`      | `cell`   ||              |            |
+| `dict`       | `struct` || `struct`     | `dict`     |
+| `str`        | `char`   || `char`       | `str`      |
+|              |          || `string`     | `str`      |
+| callable     | `handle` ||              |            |
+|              |          || `object`     | opaque     |
+| `None`       | `[]`     ||              |            |
+|numeric scalar| `double` ||numeric scalar| `np.array` |
+| `np.array`   | `array`  || `array`      | `np.array` |
+
+Note that:
+
+* Single type `list`s or `tuple`s are converted to the corresponding type Matlab `array` and _not_ `cell`.
+* Elements within containers (`list`, `tuple`, `cell`, `dict`, `struct`) are converted recursively.
+* Matlab tends to use `double` for all numeric types so any Python numeric scalars (regardless of type) are
+  converted to `double`. Because elements in containers are converted recursively, a Python `list` or `tuple`
+  of numbers (of one or mixed types) will be converted to a Matlab `double array`.
+* Matlab objects are wrapped in a Python-opaque object (unless the `call.m` gateway function is used, see below).
+  This object can be passed back to Matlab for further processing but cannot be processed in Python
+* Python objects cannot be passed to Matlab, except for callable Python classes (e.g. those with `__call__`)
+  if the `call.m` gateway function and the `call_python` mex function are used.
+* Numpy `np.array` and Matlab `array` have their underlying data wrapped in the output format.
+    + Numpy arrays support views where the data is not owned by the numpy object. `libpymcr` implements the
+      wrapping as a view with the owner being the (Python-opaque) Matlab `array` object.
+    + For the reverse (Python-to-Matlab), the Matlab API does not support views, so a hack is used 
+      as detailed below.
+    + In addition, Matlab has limited support for non-contiguous arrays, so only contiguous arrays are
+      wrapped. Non-contiguous numpy arrays are instead copied into a new Matlab `array`.
+    + Finally, whilst `libpymcr` wraps C-contiguous (row-major) arrays, many internal Matlab numeric
+      functions only work with F-contiguous (column-major) arrays so Matlab will force a data transpose copy
```

Comment:
For a 2D array, if the memory layout is contiguous (e.g. it's not a `view` or something similar) then it must be either C- (row-major) or F-contiguous (column-major). For higher dimensional arrays, the ordering can get more complicated, and is generally described by the `strides` array ([detailed explanation here](https://numpy.org/doc/stable/reference/arrays.ndarray.html#internal-memory-layout-of-an-ndarray)), although there are specific stride patterns which are used by Fortran/Matlab ("F-contiguous") for general N-D arrays, and another specific stride pattern for basic C-arrays (arrays of arrays of ... of pointers to contiguous rows).

Now, in our case, data going from Matlab to Python will always be wrapped because numpy is very flexible in how it can handle data. 

From Python to Matlab is more restrictive because Matlab only support the two specific stride patterns (C- and F-contiguous) and does not support non-contiguous arrays or general stride patterns (although I think it actually does internally, it does not expose such facilities to user code).

So, if the numpy data is a 1D vector it is always wrapped as long as the memory is actually contiguous. 

For 2D numpy arrays, `libpymcr` will always wrap the data - BUT Matlab may later cause a copy to be made for specific operations if the data was C-contiguous in Python.

For higher order ND arrays `libpymcr` will try to wrap it if the numpy array has either the `F-contiguous` or `C-contiguous` flags, otherwise it will copy the data. Internally Matlab may later copy/transpose `C-contiguous` data.